### PR TITLE
A simple outline of a meta-tracer API.

### PR DESCRIPTION
**Pull request by ltratt**
I've been chewing over https://github.com/softdevteam/yorick_jit_runtime/pull/1 and wondering if we can simplify things. Here's a possible attempt at doing so. Warning: I may have missed something important. That said, I *think* this models the same sort of thing as https://github.com/softdevteam/yorick_jit_runtime/pull/1, but with no possibility of data-races, and a simpler API.

There are a few core ideas behind this (intentionally simple API):

  1) Rather than doing magic stuff with globals, make the user responsible for creating and passing around the `MetaTracer` instance. That means that they'd need to use it in the following sort of way:

```rust
         let mt = MetaTracer::new();
         let pc = 0;
         while pc < prog.len() {
             if let ControlOutcome::Executed = mt.control_point(pc) {
                 continue;
             }
             // Normal interpreter loop here
         }
```

  2) Hide all the threading stuff about hot counts in one data-structure: as a given program location transitions from "counting" -> "tracing" -> "traced", simply map to an enum that naturally represents these 3 states.

  3) Don't expose information about the internal state that we don't know we need. Do users need to know what the JIT control_point did, beyond knowing if a trace was executed or not? I'm not sure they do. In fact, at the moment a bool would be sufficient, but I'm sort-of persuaded that an enum is easier to extend in the future if that turns out to be wrong.

### CODE REVIEW

**Comment by vext01** [comment]
Morning Laurie.

Thanks for this. I'm actually glad to throw around some new ideas. I think it's important to get this bit right now. I fear that we might "bake in" some bad design decisions and live forever with the consequences!

I've not looked at the code yet, but let me comment on a few things:

> Rather than doing magic stuff with globals, make the user responsible for creating and passing around the MetaTracer instance.

I did consider this, but favoured a more "automatic" approach for the user. Having seen the programming difficulties of this model, I am happy to place this small burden on the VM author. I'm sure they will understand :)

> Do users need to know what the JIT control_point did, beyond knowing if a trace was executed or not? I'm not sure they do.

Indeed, I did not intend to expose this at all, but I needed a race-free way for tests to know what the JIT just did in the current thread. I'm not sure if you ported over my tests, but if not, I think it would be prudent to do so. If we do need to know the last decision, perhaps we can have an "internal variant" of the control point which exposes this, then a "normal variant" which calls the internal variant, just discarding the JIT decision.

I'll comment on the code itself shortly.

**Comment by ltratt** [comment]
IIRC, you can access "private" elements within the same module's tests, so you don't need to expose things in this way.

**Comment by vext01** [comment]
It's the "race free" part that bothers me. Another thread may have called the control point between us calling and checking the state.

**Comment by vext01** [comment]
Let me check your tests before we continue this convo.

File: .travis.yml
Comment (vext01) [inline .travis.yml:2]: So presumably we don't need the threading hack here any more since there is no global state, correct? If so, I like it :)

File: Cargo.toml
Comment (vext01) [inline Cargo.toml]: Can we add me as well? I know this is all your code, but I guess we required my (failed) attempt to arrive here.

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: Let's add an XXX that this should be configurable.

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: Perhaps a more formal name would be good here? How about `InterpretedLocation`?

Comment (vext01) [reply src/lib.rs]: Or just `Interpreted`?

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: Shall we alias a type for the counter, like before?

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: I like the fact that the counter disappears when the mode changes. Nice.

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: I think this needs a different name, as it's only a counter in one of the three variants. Before I used `Phase`. How about that?

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: This is the one shared among threads? Lets say that.

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: Is `Mutex` what we want (it may well be, but I want to ask)? This means only one thread can ever observe the data. Before we had `RwLock` which would allow multiple readers.

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: What's this for?

Comment (vext01) [reply src/lib.rs]: (and yes, i read the docs ;) )

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: I think this enum needs a few more variants so that tests can check more aspects:
 * `StartedTracing`
 * `FinishedTracing`

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: That can be just `use ::*`

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: Cool, one of my outstanding questions was "Should we use Arc so that no one thread 'owns' the JIT".

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: Can we re-introduce my named-thread utility? This means the threads can be given useful names which appear in backtraces and help debugging.

https://github.com/softdevteam/yorick_jit_runtime/blob/047f3204b59507f98738a55f03a529fe98865ce8/src/lib.rs#L287-L297

**Review by vext01 — CHANGES_REQUESTED** [review]
OK, this looks like a decent start.

There's a lot of comments, I know, but let me say that I do prefer your fundamental approach to mine. It's simpler, and the concurrency is easier to reason about. I live and learn.

Aside from what I have already commented on:

 * Compilation threads are not yet modelled (and will introduce non-determinism which may highlight concurrency issues). We could do this in a separate PR (and with a thread pool from the outset?)

 * The tests are quite simple. We should try some variations of the multi-threaded test to try and force out some concurrency bugs.

Cheers

File: Cargo.toml
Comment (ltratt) [reply Cargo.toml]: In retrospect, we should probably do `Software Development Team <info@soft-dev.org>`?

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: That would be overkill. Of course it's configurable (and, one day, it'll turn into a dynamically adjustable variable).

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: I don't want to expose that to users until we have a concrete use case for doing so. Leaking so many internal details to the outside world is dangerous IMHO.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: Good idea.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: In general I'm not a big fan of type aliases because they don't imply any type constraints. [Two types `T1` and `T2` both aliased to `u16` are considered entirely equivalent in Rust.] That said, if we need to hide the details of a type, aliasing can be an easy hack. Since this is an internal detail, I'm not sure it's worth the extra effort.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: Probably a good idea. Let me cogitate a little bit.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: You can do what you want with it: there's no need to document one of the things you could do with it.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: No, `Mutex` is not a good long-term solution -- but `RwLock` also sucks as it's still a heavyweight solution for something where performance will be critical. But this is a detail for another PR. I have ideas, although they're not completely implemented yet.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: Good point -- it's not needed in this PR any more. [Sorry it's a hangover from an earlier thing.]

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: This is a test, so I don't think it's relevant here. [It's a test that should not fail; if it does, naming threads is not the main concern!]

File: Cargo.toml
Comment (vext01) [reply Cargo.toml]: If you wish, but I'd prefer the individual members to be listed, as team membership is in flux, and I'm sure people would appreciate credit for past work after they have moved on.

Comment (ltratt) [reply Cargo.toml]: We don't do it in the copyright notices, so it seems a bit inconsistent to me. But I can be persuaded.

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: The bit in the brackets is what I was suggesting. A constant is not configurable :)

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: So why does the user need to know if a trace executed then?

I think we only return these things for testing purposes.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: Fair enough, but I really don't think this is an "XXX" place. "XXX" to me means "big danger ahead if you don't consider ABC".

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: I see. Well, if there is a way to have a type-distinct counter type, I'd be all for it.

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: Good point.

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: ok, let's raise issues for these things once we merge this.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: I originally thought (as in the PR message at the top) that we could use the return type to remove some of the "magic" around control points. Alas, I was wrong. I now think this should just be a `void` return type, because if you think about things like black holing, it can potentially return control at very random points.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: There are ways, but they're heavyweight. They're only worth doing if you're worried that you will accidentally coerce between two types you would like to be distinct. That isn't a real worry here.

File: Cargo.toml
Comment (vext01) [reply Cargo.toml]: Hrm, let's think about this. I think you can see my point though?

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: Give me a chance to finish off the next PR after this one -- or to find that I'm stuck :)

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: `XXX` is the same as `TODO`. Just something to implement later.

Let's use issues instead then.

File: Cargo.toml
Comment (ltratt) [reply Cargo.toml]: Yes, but then we get to the question of "what constitutes an author". I think this field is a bit of a minefield :/

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: So my next question is, how do you test this code with a void return type?

Comment (ltratt) [reply src/lib.rs]: It may make it harder: but leaking details in a public API solely for internal testing is a recipe for disaster. I think it's paramount to remember that users will come to rely on all and any details in a public API, and do things based on that which we would really prefer them not to.

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: Fine, let's leave it for now.

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: How about some kind of JIT log, which details the work of each thread, that can be parsed later?

Comment (vext01) [reply src/lib.rs]: Maybe not parsed, it could be in-memory.

Comment (ltratt) [reply src/lib.rs]: Maybe: but that's way too much for this PR, and will require some careful thought and experimentation (including, I suspect, some performance measurements).

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: I disagree on this one. If there is a concurrency bug causing a non-deterministic crash, we may not be able to easily reproduce it, and I'd like all the info I can get. Without my helper, all the test threads will have the same name.

Comment (ltratt) [reply src/lib.rs]: This is over-engineering in a test that isn't expected to fail. Why not preemptively add extra information that might be useful if a test fails? I could put a print statement between every line, for example. [Yes, that's a slightly facetious example, but there is a serious point here.]

**Comment by ltratt** [comment]
OK, see if those last two commits make sense or not and whether they're complete or not -- let me know if they don't!

**Comment by ltratt** [comment]
Two sort-of-meta questions (but we might as well think about these now):

1) The name of this repo and library. `yorick_jit_runtime` is just too bulky IMHO, especially given that this is the user-facing API. Without wishing to disappear *too* far down the naming-black-hole, I suggest we name the library `yorick`; and, at least for the time being, maybe call the repo `yorickrt`? I'd like to think that ultimately we might be able to do nearly everything yorick-related in one repo (with different Rust sub-projects), though the rustc shenanigans will, I think, inevitably muddy the waters.

2) Sticking everything in `lib.rs` isn't sustainable: we're clearly going to need to split things into modules for our own sanity. As a first step, I suggest that I make `src/metatracer.rs` and move what are currently the contents of `lib.rs` into there.

Any thoughts?

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: [My earlier explanation was a bit terse. I should have said that one problem with `RwLock` is that it doesn't provide an "upgrade" from read to write locks. Indeed, it's impossible to guarantee that any read lock can be upgraded (since two readers could request an upgrade simultaneously). So it's difficult to make a nice API for this. However, the bigger problem is that even the read lock is going to be too expensive for us in the long run, since *mostly* in a long-running program we expect to say "do you have a trace for location L" and the answer to be "yes". i.e. for long-running programs most of the time is going to be spent returning the same answer. So we probably want to use a (lightweight) atomic read rather than a heavy-weight lock. I've sketched a couple of solutions, but have a better one in mind. It'll take a few days.]

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: When I was working on #1, often test failures could only occurred on travis, and I found myself having to push debugging code, then wait for travis to crash (possibly re-starting the tests several times to reproduce the crash due to non-determinism), then repeat.

Rather than go through this dance every time, I think it would save us time to have diagnostics in place (as long as performance isn't impacted -- thread names should be fine). After all, it may be *users* finding the crashes. Do we want to ask them to apply patches and re-compile? Or would we rather they just post a back trace?

I'd suggest we name all our threads, in tests or otherwise.

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: So this is an abbreviation of "Location Phase"? Does a location have a phase? Or is the phase a property of the thread?

**Comment by vext01** [comment]
The commits you pushed are fine with one minor comment about naming.

1) OK with me. Go ahead.

2) `metatracer.rs` is a pretty generic name. Would it help to think about what other functionality you expect to appear in this library? I guess to make a palatable API, much of the stuff under your proposed sub-module will need to be exposed to the top-level of the crate (e.g. we probably don't want `yorick::metatracer::jit_control_point`, but rather `yorick::jit_control_point`?). I guess what I'm saying is that the API is *somewhat* (but not totally) independent of the internal file structure.

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: Understood about upgrading the locks. I had that very same conversation with someone last week!

But note that in the utopian circumstance:

> mostly in a long-running program we expect to say "do you have a trace for location L" and the answer to be "yes"

A `RwLock` won't be terrible, as you will only have readers.

For me the shortcoming of the `RwLock` is where a new trace needs to be built and in order to insert the trace, we would have to wait for N readers to release their read locks.

I'm not saying `RwLock` is the best solution, but I do think it's probably better than a `Mutex` in terms of lock contention for our "expected common case" (however, I always expect the unexpected since the warmup experiment!).

Anyway, I await your new proposals, good luck! :)

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: this can be:
```
LocPhase::Tracing =>  v.insert(LocPhase::Trace),
```

To save some vertical space.

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: Similarly here, we can save some vertical space.

**Comment by vext01** [comment]
I can't seem to find the comment now, but we were talking about whether or not we should be able to detect if the JIT started/stopped tracing after a given invocation of the control point. In your new testing strategy (after killing the return type of the control point), it might be worth thinking about those.

Also about how to test once compilation happens in a thread. Then you have latency between starting compiling a thread, and the trace appearing, ready for use (this is why we had `Pending` and `Ready` trace types in #1. Tests needed a loop with a timeout to decide if compilation was working or not).

**Comment by vext01** [comment]
To expand more concretely, although you have `TraceLoc::Tracing` and `TraceLoc::Trace`, should there not be one more state which indicates that we are not tracing the location (we finished tracing), but the trace is not yet ready (is being compiled in the background). I think the distinction will be needed to prevent another thread tracing the same location.

File: src/lib.rs
Comment (vext01) [inline src/lib.rs]: Something doesn't sit right with me about `Tracing`/`Trace`. These are pretty similarly named.

How about `BeingTraced` and `TraceReady`? (and eventually `TraceCompiling`? see comments below about this state)

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: I don't think we're going to agree on this one -- sorry.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: Yes, yes, no (respectively) :)

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: Unfortunately, I don't think so. Each arm needs to return the type `()`, so we have to use curly brackets so that we can use `;`, forcing the return type to be `()`... I assume this is all due to limitations in the parser.

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: I'll try and put some numbers on this soon, as my current experiment wasn't exactly rigorous ;)

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: So you don't think it would be useful to have threads named `compiling_loc_45` for example?

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: ok

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: How about: Counting, Tracing, Compiled?

In my head, at the moment, I'm including everything about compilation under "tracing". Put another way, "tracing" means "don't start another trace". Maybe we will need another state, but perhaps it's better to wait until we're sure?

File: src/lib.rs
Comment (ltratt) [reply src/lib.rs]: I think we should do that for user visible threads. But there's no suggestion of a race in this test, so the extra support is bloat. And, the older I get, the more I think that bloat in code is our eternal enemy (literally: bloat never dies).

**Comment by ltratt** [comment]
In terms of the high-level API, I'm imagining people doing `yorickrt::MetaTracer::new(...)` or similar. Or maybe just `yorickrt::metatracer(...)`. Any preferences?

File: src/lib.rs
Comment (vext01) [reply src/lib.rs]: The names are perfect, let's use those.

I think without the distinction you are likely to spawn duplicate compilation threads. It works now because traces are assumed to appear synchronously.

It's up to you whether you introduce the new state now or in the PR that introduces compilation threads.

**Comment by vext01** [comment]
The former makes more sense to me.

**Comment by ltratt** [comment]
OK, I think those commits reflect our conversation this morning. Comments welcome!

**Comment by vext01** [comment]
I think this is a decent start point that we can then iterate upon.

I have a gut feeling that a lot of our initial design decisions will change as we add more complexity, but that's fine, right? If we try an crowbar all of the functionality in at once, we will end up with huge PRs.

So, what's the plan here? Squash an merge soon?

**Comment by ltratt** [comment]
I think this entire line of work is going to stretch our programming abilities to the maximum. We're going to have to focus on simplicity, so that as things change, we don't have to do any more reengineering than we absolutely have to.

I'll squash now...

**Comment by ltratt** [comment]
OK, squashed. [Note that I've headily edited the first commit message to reflect all the changes we've made. Please check!]

**Review by vext01 — APPROVED** [review]

**Comment by vext01** [comment]
LGTM.

**Pull request merged**
